gobblin-runtime/src/main/java/org/apache/gobblin/quality/DataQualityEvaluator.java

```diff
@@ -0,0 +1,191 @@
+     *
+     * @param taskStates List of task states to evaluate
+     * @param jobState The job state containing additional context
+     * @return DataQualityEvaluationResult containing the evaluation results
+     */
+    public static DataQualityEvaluationResult evaluateDataQuality(List<TaskState> taskStates, JobState jobState) {
+        DataQualityStatus jobDataQuality = DataQualityStatus.PASSED;
+        int totalFiles = 0;
+        int failedFilesSize = 0;
+        int passedFilesSize = 0;
+        int nonEvaluatedFilesSize = 0;
+
+        for (TaskState taskState : taskStates) {
+            totalFiles++;
+
+            // Handle null task states gracefully
+            if (taskState == null) {
+                log.warn("Encountered null task state, skipping data quality evaluation for this task");
+                nonEvaluatedFilesSize++;
+                continue;
+            }
+
+            DataQualityStatus taskDataQuality = null;
+            String result = taskState.getProp(ConfigurationKeys.TASK_LEVEL_POLICY_RESULT_KEY);
+            taskDataQuality = DataQualityStatus.fromString(result);
+            if (taskDataQuality != DataQualityStatus.NOT_EVALUATED) {
+                log.debug("Data quality status of this task is: " + taskDataQuality);
+                if (DataQualityStatus.PASSED == taskDataQuality) {
+                    passedFilesSize++;
+                } else if (DataQualityStatus.FAILED == taskDataQuality){
+                    failedFilesSize++;
+                    jobDataQuality = DataQualityStatus.FAILED;
+                }
+            } else {
+                // Handle files without data quality evaluation
+                nonEvaluatedFilesSize++;
+                log.warn("No data quality evaluation for task: " + taskState.getTaskId());
+            }
+        }
+
+        // Log summary of evaluation
+        log.info("Data quality evaluation summary - Total: {}, Passed: {}, Failed: {}, Not Evaluated: {}",
+            totalFiles, passedFilesSize, failedFilesSize, nonEvaluatedFilesSize);
+        return new DataQualityEvaluationResult(jobDataQuality, totalFiles, passedFilesSize, failedFilesSize, nonEvaluatedFilesSize);
+    }
+
+    private static void emitMetrics(JobState jobState, int jobDataQuality, int totalFiles,
+            int passedFilesSize, int failedFilesSize, int nonEvaluatedFilesSize, String datasetUrn) {
+        OpenTelemetryMetricsBase otelMetrics = OpenTelemetryMetrics.getInstance(jobState);
+        if(otelMetrics != null) {
+            Meter meter = otelMetrics.getMeter(GAAS_OBSERVABILITY_METRICS_GROUPNAME);
+            AtomicLong jobDataQualityRef = new AtomicLong(jobDataQuality);
+            AtomicLong totalFilesRef = new AtomicLong(totalFiles);
+            AtomicLong passedFilesRef = new AtomicLong(passedFilesSize);
+            AtomicLong failedFilesRef = new AtomicLong(failedFilesSize);
+            AtomicLong nonEvaluatedFilesRef = new AtomicLong(nonEvaluatedFilesSize);
```

Comment:
Any reason to use `AtomicLong`? It should not be needed since the metric values are immutable snapshots and not being mutated or shared across threads